### PR TITLE
fix: correct ordering of accessibility traits

### LINKED ISSUE #1131: Searching by accessibility traits does not work correctly.
#### Description
When searching for an element with the trait `header` I noticed that the Element matcher was searching for the trait `UIAccessibilityTraitImage`. 

It looks like the flags are set incorrectly in `src/ios/earlgreyapi/GREYMatchers.js` (https://github.com/wix/Detox/blob/master/detox/src/ios/earlgreyapi/GREYMatchers.js).

I found the following list which seems pretty accurate:
https://docs.microsoft.com/en-us/dotnet/api/UIKit.UIAccessibilityTrait

I'd be happy to submit a PR for this but I'd appreciate some guidance as to whether this change would have any knock-on impact (I assume the list was correct at one point) and given that [accessibility traits are being deprecated](https://docs.microsoft.com/en-us/dotnet/api/UIKit.UIAccessibilityTrait) whether it is better to leave it.

#### Steps to Reproduce

Search for an element by any trait other than `button` and `link` and you will see it is mismatched.

#### Detox, Node, Device, Xcode and macOS Versions

* Detox: 9.1.2
* React Native: 0.57.8
* Node: 8.12.0
* Device: iPhone X (iOS 12.1)
* Xcode: 10.1
* macOS: Mojave 10.14.3

#### Device and verbose Detox logs

I don't believe the logs add value for this issue but I can provide them if required.

Here is a snippet from the error trace for the following test:

```javascript
it("should show the client's name", async () => {
  await expect(
    element(by.text("Mr C R Green").and(by.traits(["header"])))
  ).toBeVisible()
})
```

    Failed: [Error: Error: Cannot find UI Element.
    Exception with Assertion: {
      "Assertion Criteria":  "assertWithMatcher:matcherForSufficientlyVisible(>=0.750000)",
      "Element Matcher":  "((!(kindOfClass('RCTScrollView')) && ((((kindOfClass('UILabel') || kindOfClass('UITextField') || kindOfClass('UITextView')) && hasText('Mr C R Green')) || (kindOfClass('RCTTextView') && an object with accessibilityLabel "Mr C R Green")) && ((respondsToSelector(isAccessibilityElement) && isAccessibilityElement) && accessibilityTraits: UIAccessibilityTraitImage))) || (((kindOfClass('UIView') || respondsToSelector(accessibilityContainer)) && parentThatMatches(kindOfClass('RCTScrollView'))) && ((kindOfClass('UIView') || respondsToSelector(accessibilityContainer)) && parentThatMatches(((((kindOfClass('UILabel') || kindOfClass('UITextField') || kindOfClass('UITextView')) && hasText('Mr C R Green')) || (kindOfClass('RCTTextView') && an object with accessibilityLabel "Mr C R Green")) && ((respondsToSelector(isAccessibilityElement) && isAccessibilityElement) && accessibilityTraits: UIAccessibilityTraitImage))))))",
      "Recovery Suggestion":  "Check if the element exists in the UI hierarchy printed below. If it exists, adjust the matcher so that it accurately matches element."
    }

    Error Trace: [
      {
        "Description":  "Interaction cannot continue because the desired element was not found.",
        "Error Domain":  "com.google.earlgrey.ElementInteractionErrorDomain",
        "Error Code":  "0",
        "File Name":  "GREYElementInteraction.m",
        "Function Name":  "-[GREYElementInteraction matchedElementsWithTimeout:error:]",
        "Line":  "124"
      }
    ]

### PR DESCRIPTION
**Pull request by ximenean**
- [x] This is a small change. 
---
Fixes #1131

**Description:**

When searching for an element with the trait header I noticed that the Element matcher was searching for the trait UIAccessibilityTraitImage.

Corrected using the following list:
docs.microsoft.com/en-us/dotnet/api/UIKit.UIAccessibilityTrait

### CODE REVIEW

**Comment by LeoNatan** [comment]
Hello,

Where do these values come from? Is it a translation from a native enum?

Thanks

**Comment by ximenean** [comment]
Hi. TBH I had difficulty tracking it down as I've not worked with native code. I found this list:
https://docs.microsoft.com/en-us/dotnet/api/UIKit.UIAccessibilityTrait

and verified it based on the log output, for example if I set a trait of 'header' I would see it come out as `accessibilityTraits: UIAccessibilityTraitImage)`

There are some more details in the associated issue that might be helpful https://github.com/wix/Detox/issues/1131

**Comment by LeoNatan** [comment]
The issue explains the problem, but does not help understanding your changes here. I don't know what that link is, as it's Xamarin code. Please verify with the iOS SDK and explain how each value is calculated. Thanks

File: detox/test/e2e/02.matchers.test.js [hunk 1]
```diff
@@ -4,19 +4,21 @@ describe('Matchers', () => {
...
+  it('should match elements by (accessibility) id', async () => {
     await element(by.id('UniqueId345')).tap();
     await expect(element(by.text('ID Working!!!'))).toBeVisible();
   });
 
   it('should match elements by index', async () => {
     const index = device.getPlatform() === 'ios' ? 2 : 0;
-    await element(by.text('Index')).atIndex(index).tap();
+    await element(by.text('Index'))
+      .atIndex(index)
```
Comment (LeoNatan) [inline detox/test/e2e/02.matchers.test.js]: Any reason to break this into multiple lines? Please revert this change.

File: detox/test/e2e/02.matchers.test.js [hunk 2]
```diff
@@ -30,8 +32,8 @@ describe('Matchers', () => {
 
   // https://facebook.github.io/react-native/docs/accessibility.html#accessibilitytraits-ios
   // Accessibility Inspector in the simulator can help investigate traits
-  it(':ios: should match elements by accesibility trait', async () => {
-    await element(by.traits(['button', 'text'])).tap();
+  it(':ios: should match elements by accessibility trait', async () => {
+    await element(by.traits(['button', 'disabled'])).tap();
```
Comment (LeoNatan) [inline detox/test/e2e/02.matchers.test.js]: I don't understand this change. If the traits are a button and disabled, how can you tap it? That's not how accessibility traits work.

File: detox/test/e2e/02.matchers.test.js (hunk 1 above)
Comment (ximenean) [reply detox/test/e2e/02.matchers.test.js]: That's a changed done by Prettier. I can revert it.

File: detox/test/e2e/02.matchers.test.js (hunk 2 above)
Comment (ximenean) [reply detox/test/e2e/02.matchers.test.js]: This is a straight swap for what the test was doing originally. The accessibility trait of 'text' would have actually corresponded to 'disabled'

Comment (LeoNatan) [reply detox/test/e2e/02.matchers.test.js]: I don't think this is true, because a button also has text. If an accessibility element is disabled, it should not be interactive.

**Comment by ximenean** [comment]
The link to the Xamarin enum just shows what the correct values are. 

As I say I verified the values additionally by checking against the output from Detox to see what the trait provided in the Detox test corresponded to in EarlGrey.

I will try to look at the iOS SDK to find something more convincing, but as I say, I'm not a native developer.

**Comment by LeoNatan** [comment]
Here is the output of all the available traits in the SDK:

```
2019-03-13 12:46:58.398288+0200 example[3038:54700] (
        {
        UIAccessibilityTraitNone = 0;
    },
        {
        UIAccessibilityTraitButton = 1;
    },
        {
        UIAccessibilityTraitLink = 2;
    },
        {
        UIAccessibilityTraitHeader = 65536;
    },
        {
        UIAccessibilityTraitSearchField = 1024;
    },
        {
        UIAccessibilityTraitImage = 4;
    },
        {
        UIAccessibilityTraitSelected = 8;
    },
        {
        UIAccessibilityTraitPlaysSound = 16;
    },
        {
        UIAccessibilityTraitKeyboardKey = 32;
    },
        {
        UIAccessibilityTraitStaticText = 64;
    },
        {
        UIAccessibilityTraitSummaryElement = 128;
    },
        {
        UIAccessibilityTraitNotEnabled = 256;
    },
        {
        UIAccessibilityTraitUpdatesFrequently = 512;
    },
        {
        UIAccessibilityTraitStartsMediaSession = 2048;
    },
        {
        UIAccessibilityTraitAdjustable = 4096;
    },
        {
        UIAccessibilityTraitAllowsDirectInteraction = 8192;
    },
        {
        UIAccessibilityTraitCausesPageTurn = 16384;
    },
        {
        UIAccessibilityTraitTabBar = 32768;
    }
)
```

File: detox/test/e2e/02.matchers.test.js (hunk 2 above)
Comment (ximenean) [reply detox/test/e2e/02.matchers.test.js]: OK. Well something is misaligned somewhere. If I keep the value as `text` the test correctly finds elements with a `text` trait but fails as there are more than one.

**Comment by ximenean** [comment]
Thanks. That looks to match my changes, right?

**Comment by LeoNatan** [comment]
Seems to match, yes. Do you mind adding the missing one?

File: detox/test/e2e/02.matchers.test.js (hunk 2 above)
Comment (LeoNatan) [reply detox/test/e2e/02.matchers.test.js]: https://developer.apple.com/documentation/uikit/uiaccessibilitytraitnotenabled?language=objc

> The accessibility element is not enabled and does not respond to user interaction.

File: detox/test/e2e/02.matchers.test.js (hunk 1 above)
Comment (LeoNatan) [reply detox/test/e2e/02.matchers.test.js]: Please revert. Thanks

File: detox/test/e2e/02.matchers.test.js (hunk 2 above)
Comment (ximenean) [reply detox/test/e2e/02.matchers.test.js]: I don't dispute that. But as you can see the `text` trait would correspond to the value of 256 which should be `disabled` so does that fact that this test passes suggest a problem elsewhere.

Comment (LeoNatan) [reply detox/test/e2e/02.matchers.test.js]: If you search for button+text (with the correct values), which element is selected? Likewise, if you search for button+disabled (with the correct values), which element is selected?

Comment (ximenean) [reply detox/test/e2e/02.matchers.test.js]: With button+text (corrected values) I get ```  1) Matchers
       :ios: should match elements by accessibility trait:
     Error: Error: Multiple UI elements matched for the given criteria.
Exception with Action: {
  "Action Name":  "Tap",
  "Element Matcher":  "((!(kindOfClass('RCTScrollView')) && ((respondsToSelector(isAccessibilityElement) && isAccessibilityElement) && accessibilityTraits: UIAccessibilityTraitButton,UIAccessibilityTraitStaticText)) || (((kindOfClass('UIView') || respondsToSelector(accessibilityContainer)) && parentThatMatches(kindOfClass('RCTScrollView'))) && ((kindOfClass('UIView') || respondsToSelector(accessibilityContainer)) && parentThatMatches(((respondsToSelector(isAccessibilityElement) && isAccessibilityElement) && accessibilityTraits: UIAccessibilityTraitButton,UIAccessibilityTraitStaticText)))))",
  "Recovery Suggestion":  "Create a more specific matcher to uniquely match an element. If that's not possible then use atIndex: to select from one of the matched elements but the order of elements may change."
}
```

Comment (ximenean) [reply detox/test/e2e/02.matchers.test.js]:
```
Error Trace: [
  {
    "Description":  "Multiple elements were matched: (
    "<UILabel:0x7fe33b520410; AX=Y; AX.label='Active'; AX.frame={{147.16666666666666, 26}, {80.666666666666657, 36}}; AX.activationPoint={187.5, 44}; AX.traits='UIAccessibilityTraitStaticText'; AX.focused='N'; frame={{147.16666666666666, 26}, {80.666666666666671, 36}}; opaque; alpha=1; UIE=N; text='Active'>",
    "<RCTView:0x7fe33d9350b0; AX=Y; AX.label='Traits'; AX.frame={{169.33332824707031, 288.33334350585938}, {36.333343505859375, 27}}; AX.activationPoint={187.5, 301.83334350585938}; AX.traits='UIAccessibilityTraitButton,UIAccessibilityTraitStaticText'; AX.focused='N'; frame={{169.33332824707031, 288.33334350585938}, {36.333343505859375, 27}}; opaque; alpha=1>",
    "<RCTTextView:0x7fe33d938d90; AX=Y; AX.label='Index'; AX.frame={{170, 580.66668701171875}, {35.666656494140625, 17.3333740234375}}; AX.activationPoint={187.83332824707031, 589.3333740234375}; AX.traits='UIAccessibilityTraitStaticText'; AX.focused='N'; frame={{0, 0}, {35.666656494140625, 17.3333740234375}}; alpha=1>",
    "<RCTTextView:0x7fe33d938680; AX=Y; AX.label='Index'; AX.frame={{170, 543.66668701171875}, {35.666656494140625, 17.3333740234375}}; AX.activationPoint={187.83332824707031, 552.3333740234375}; AX.traits='UIAccessibilityTraitStaticText'; AX.focused='N'; frame={{0, 0}, {35.666656494140625, 17.3333740234375}}; alpha=1>",
    "<RCTTextView:0x7fe33d937f70; AX=Y; AX.label='Index'; AX.frame={{170, 506.66665649414062}, {35.666656494140625, 17.333343505859375}}; AX.activationPoint={187.83332824707031, 515.33332824707031}; AX.traits='UIAccessibilityTraitStaticText'; AX.focused='N'; frame={{0, 0}, {35.666656494140625, 17.333343505859375}}; alpha=1>",
    "<RCTTextView:0x7fe33d937860; AX=Y; AX.id='ProductId003'; AX.label='Product'; AX.frame={{268.66666793823242, 459.33334350585938}, {51.66668701171875, 17.33331298828125}}; AX.activationPoint={294.5000114440918, 468}; AX.traits='UIAccessibilityTraitStaticText'; AX.focused='N'; frame={{224, 10}, {51.66668701171875, 17.33331298828125}}; alpha=1>",
    "<RCTTextView:0x7fe33d937450; AX=Y; AX.id='ProductId002'; AX.label='Product'; AX.frame={{197.33333969116211, 459.33334350585938}, {51.666671752929688, 17.33331298828125}}; AX.activationPoint={223.16667556762695, 468}; AX.traits='UIAccessibilityTraitStaticText'; AX.focused='N'; frame={{152.66667175292969, 10}, {51.666671752929688, 17.33331298828125}}; alpha=1>",
    "<RCTTextView:0x7fe33d937040; AX=Y; AX.id='ProductId001'; AX.label='Product'; AX.frame={{126.00000381469727, 459.33334350585938}, {51.666671752929688, 17.33331298828125}}; AX.activationPoint={151.83333969116211, 468}; AX.traits='UIAccessibilityTraitStaticText'; AX.focused='N'; frame={{81.333335876464844, 10}, {51.666671752929688, 17.33331298828125}}; alpha=1>",
    "<RCTTextView:0x7fe33d936c30; AX=Y; AX.id='ProductId000'; AX.label='Product'; AX.frame={{54.666667938232422, 459.33334350585938}, {51.666667938232422, 17.33331298828125}}; AX.activationPoint={80.500001907348633, 468}; AX.traits='UIAccessibilityTraitStaticText'; AX.focused='N'; frame={{10, 10}, {51.666667938232422, 17.33331298828125}}; alpha=1>",
    "<RCTTextView:0x7fe33d934960; AX=Y; AX.label='Traits'; AX.frame={{169.33332824707031, 288.33334350585938}, {36.333343505859375, 17.33331298828125}}; AX.activationPoint={187.5, 297}; AX.traits='UIAccessibilityTraitStaticText'; AX.focused='N'; frame={{0, 0}, {36.333343505859375, 17.33331298828125}}; alpha=1>",
    "<RCTTextView:0x7fe33d924140; AX=Y; AX.id='UniqueId345'; AX.label='ID'; AX.frame={{180.66667175292969, 251.33332824707031}, {14.333343505859375, 17.333328247070312}}; AX.activationPoint={187.83334350585938, 259.99999237060547}; AX.traits='UIAccessibilityTraitStaticText'; AX.focused='N'; frame={{0, 0}, {14.333343505859375, 17.333328247070312}}; alpha=1>",
    "<RCTTextView:0x7fe33d91e1a0; AX=Y; AX.label='Label'; AX.frame={{169.66667175292969, 214.33332824707031}, {35.666656494140625, 17.333343505859375}}; AX.activationPoint={187.5, 223}; AX.traits='UIAccessibilityTraitStaticText'; AX.focused='N'; frame={{0, 0}, {35.666656494140625, 17.333343505859375}}; alpha=1>"
). Please use selection matchers to narrow the selection down to single element.",
    "Error Domain":  "com.google.earlgrey.ElementInteractionErrorDomain",
    "Error Code":  "5",
    "File Name":  "GREYElementInteraction.m",
    "Function Name":  "-[GREYElementInteraction grey_errorForMultipleMatchingElements:withMatchedElementsIndexOutOfBounds:]",
    "Line":  "956"
  }
```

Comment (LeoNatan) [reply detox/test/e2e/02.matchers.test.js]: I don't understand what is going on here. Only one of these is a button in addition to text: `RCTView:0x7fe33d9350b0`. I don't understand why it finds the rest.

Comment (LeoNatan) [reply detox/test/e2e/02.matchers.test.js]: And none of them are disabled.

Comment (LeoNatan) [reply detox/test/e2e/02.matchers.test.js]: OK, I see the problem. User Accessibility Inspector to check out all the buttons there, and indeed they are all text+button. So it finds them all.

Could you please disable the test? It's a bad test. Thanks

Comment (ximenean) [reply detox/test/e2e/02.matchers.test.js]: Do you mean disable (with `it.skip`) or delete?

Comment (LeoNatan) [reply detox/test/e2e/02.matchers.test.js]: Skip. Hopefully, we'll look at it eventually and make it a test that makes sense.

**Comment by LeoNatan** [comment]
Cool, thanks! Let's wait for the tests to finish and I'll merge.

**Comment by LeoNatan** [comment]
Merged, thanks!
